### PR TITLE
Zaborins

### PR DESCRIPTION
**Pull request by 33ta**
initial database module implementation

### CODE REVIEW

File: Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        return localInstance;
+    }
+
+    /**
+     * Этот метод проверяет базу данных на пригодность
+     *
+     * @return true, если установлено соединение с корректным файлом, иначе - false
+     * @throws SQLException, если произошла ошибка соединения
+     */
+    private boolean checkDatabaseFile(Connection conn) throws SQLException {
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java]: If this is utility method, probably should make it safe by catching all nested exceptions and returning back `false`

File: Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        Connection connection = connectionsHolder.getReadingConnection();
+
+
+        try (Statement stmt = connection.createStatement()) {
+            logger.debug("reading...");
+            stmt.execute("select * from users;");
+        }
+
+
+        connection.close();
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java]: Read more about try-with-resources

File: Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    /**
+     * Этот метод - просто пример, шаблон для будущих методов записи в БД
+     *
+     * @throws SQLException если запись в бд закончилась неудачей
+     */
+    public void writeToDB() throws SQLException {
+        synchronized (this) {
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java]: Making synchronized block inside method is generally slower, better define the whole method as `synchronized`

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    private static final Logger logger = Logger.getLogger(ChatDatabaseHandler.class);
+    /**
+     * Поле, содержащее размер "пула" соединний
+     */
+    private static final int POOL_SIZE = 5;
+    /**
+     * Поле, содержащее соединения, предназначенные для чтения и не используемые в данный момент
+     */
+    private final BlockingQueue<SQLitePooledConnection> readingConnections =
+            new ArrayBlockingQueue<>(POOL_SIZE);
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: Are you sure about your collection choice? Have you read this article: http://tutorials.jenkov.com/java-util-concurrent/blockingqueue.html?

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    private final SQLitePooledConnection writingConnection;
+
+    /**
+     * Конструктор, который инициализирует соединения с базой данных
+     *
+     * @param dataSource база данных, с которой устанавливается соединение
+     * @throws SQLException, если не удалось инстанцировать объект этого класса
+     */
+    ConnectionsHolder(SQLiteConnectionPoolDataSource dataSource) throws SQLException {
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: Access modifier?

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        try {
+            ReadingConnectionListener readingConnectionListener = new ReadingConnectionListener();
+
+            while (readingConnections.remainingCapacity() != 0) {
+                SQLitePooledConnection conn =
+                        (SQLitePooledConnection) dataSource.getPooledConnection();
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: Have you read this answer: http://stackoverflow.com/a/15822838?

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            }
+        }
+
+        /**
+         * Метод, который реагирует на ошибки соединений
+         * @param event параметр, описывающий событие
+         */
+        @Override
+        public void connectionErrorOccurred(ConnectionEvent event) {
+            throw new RuntimeException("Connection error occurred");
```
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: Wow, that's a very bad practice, remove it and / or add a workaround.

File: Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    /**
+     * Статический метод, который должен быть использован при желании получить экземпляр этого
+     * класса
+     *
+     * @return единственный экземпля этого класса
+     * @throws SQLException если не удалось создать экземпляр класса
+     */
+    public static ChatDatabaseHandler getInstance() throws SQLException {
```
Comment (33ta) [inline Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java]: The problem with initialization-on-demand holder is that in my implementation constructor throws an exception

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java (hunk 5 above)
Comment (33ta) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: I thought that package-private access is okay here

File: Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java (hunk 8 above)
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ChatDatabaseHandler.java]: You can always create an "exceptionless" workaround

File: Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java (hunk 5 above)
Comment (SuppieRK) [inline Project2/src/main/java/org/nnstu2/database/ConnectionsHolder.java]: Package-private is not okay almost everywhere.

**Pull request closed without merging**